pynestml/codegeneration/nest_code_generator_utils.py

```diff
@@ -19,36 +19,119 @@
-from pynestml.symbols.variable_symbol import BlockType
+from typing import List, Optional
+
+import re
+import uuid
+
+from pynestml.frontend.pynestml_frontend import generate_nest_target
 from pynestml.meta_model.ast_variable import ASTVariable
+from pynestml.symbols.variable_symbol import BlockType
+from pynestml.symbols.variable_symbol import VariableSymbol
 
 
 class NESTCodeGeneratorUtils:
 
     @classmethod
     def print_symbol_origin(cls, variable_symbol: VariableSymbol, variable: ASTVariable) -> str:
-        """
+        r"""
         Returns a prefix corresponding to the origin of the variable symbol.
         :param variable_symbol: a single variable symbol.
         :return: the corresponding prefix
         """
         if variable_symbol.block_type in [BlockType.STATE, BlockType.EQUATION]:
             if "_is_numeric" in dir(variable) and variable._is_numeric:
-                return 'S_.ode_state[State_::%s]'
+                return "S_.ode_state[State_::%s]"
 
-            return 'S_.%s'
+            return "S_.%s"
 
         if variable_symbol.block_type == BlockType.PARAMETERS:
-            return 'P_.%s'
+            return "P_.%s"
 
         if variable_symbol.block_type == BlockType.COMMON_PARAMETERS:
-            return 'cp.%s'
+            return "cp.%s"
 
         if variable_symbol.block_type == BlockType.INTERNALS:
-            return 'V_.%s'
+            return "V_.%s"
 
         if variable_symbol.block_type == BlockType.INPUT:
-            return 'B_.%s'
+            return "B_.%s"
+
+        return ""
+
+    @classmethod
+    def generate_code_for(cls,
+                          nestml_neuron_model: str,
+                          nestml_synapse_model: Optional[str] = None,
+                          post_ports: Optional[List[str]] = None,
+                          mod_ports: Optional[List[str]] = None,
+                          uniq_id: Optional[str] = None,
+                          logging_level: str = "WARNING"):
+        """Generate code for a given neuron and synapse model, passed as a string.
+        NEST cannot yet unload or reload modules. This function implements a workaround using UUIDs to generate unique names.
+        The neuron and synapse models can be passed directly as strings in NESTML syntax, or as filenames, in which case the NESTML model is loaded from the given filename.
+
+        Returns
+        -------
+        If a synapse is specified, returns a tuple (module_name, mangled_neuron_name, mangled_synapse_name) containing the names that can be used in ``nest.Install()``, ``nest.Create()`` and ``nest.Connect()`` calls. If no synapse is specified, returns a tuple (module_name, mangled_neuron_name).
+        """
+
+        # generate unique ID
+        if uniq_id is None:
+            uniq_id = str(uuid.uuid4().hex)
+
+        # read neuron model from file?
+        if not "\n" in nestml_neuron_model and ".nestml" in nestml_neuron_model:
```

Comment:
Why is there a check for `\n` here?

Reply:
It's a heuristic check to see if we are being passed a file name, or a complete NESTML model as one big long string.